include/benchmark/benchmark_api.h

```diff
@@ -239,6 +240,149 @@ inline BENCHMARK_ALWAYS_INLINE void DoNotOptimize(Tp const& value) {
 // FIXME Add ClobberMemory() for non-gnu compilers
 #endif
 
+
+
+class Counter {
+public:
+
+  typedef enum {
+    kDefaults       = 0x00,
+    /** Mark the counter as a rate. It will be presented divided by the duration of the benchmark. */
+    kIsRate         = 0x01 << 0,
+    /** Mark the counter as a thread-average quantity. It will be presented divided by the number of threads. */
+    kAvgThreads     = 0x01 << 1
+  } Flags;
+
+public:
+
+  Counter();
+  Counter(const char *name);
+  Counter(const char *name, double v);
+  Counter(const char *name, double v, uint32_t flags);
+  ~Counter();
+  Counter(Counter const& that);
+  Counter& operator= (Counter const& that);
+#ifdef BENCHMARK_HAS_CXX11
+  Counter(Counter && that);
+  Counter& operator= (Counter && that);
+#endif
+
+  void Set(double v) { value_ = v; }
+
+  void Finish(double cpu_time, double num_threads);
+
+  const char* Name() const { return name_; }
+  double Value() const { return value_; }
+
+  bool SameName(const char *n) const;
+
+  Counter& operator = (double v) { value_  = v; return *this; }
+  Counter& operator+= (double v) { value_ += v; return *this; }
+  Counter& operator-= (double v) { value_ -= v; return *this; }
+  Counter& operator*= (double v) { value_ *= v; return *this; }
+  Counter& operator/= (double v) { value_ /= v; return *this; }
+
+private:
+
+  // using raw C strings here to avoid inclusion of std::string in the public api
+  char*    name_;     // this points to either name_buf_ or name_mem_
+  double   value_;
+  uint32_t flags_;
+
+  char name_buf_[16]; // names smaller than 15 characters are stored here
+  char* name_mem_;    // whereas longer strings are stored here
+  size_t mem_size_;   // current size of name_mem_
+
+  void SetName_(const char *name);
+};
+
+
+
+class BenchmarkCounters {
```

Comment:
I have seen the benchmarks, but I remain unconvinced by the wisdom of having a half-baked map-like class that's not quite like a map.  In particular (1) the iterator exposes internal details (such as the fact that the element is not a pair) and (2) the method names are misleading wrt their complexity (Find is O(N)).

I don't quite buy the argument that this class has to be super-efficient and, like @EricWF, I would prefer to use STL containers whenever possible.  Your benchmark is a bit of a red-herring: anyone wanting to update the counters in the tight loop would be well-advised to grab an iterator to the counter before the loop and use it in the loop.  A `std::map` would work just fine for that purpose.

As it stands this PR increases considerably the surface of the API.  While I think that user-defined counters are a great idea, there has to be a way to add them with a much more compact API, and without an implementation that does its own management of storage and containers.